File: vm/format/java/src/main/java/dev/argon/argonvm/VMClass.java
package dev.argon.argonvm;

import java.util.List;
import java.util.Map;

public interface VMClass {
	VMClass baseClass();
	List<VMType> fields();
	List<SlotSignature> slots();
	long getImplementationFunctionId(SlotKey slotKey);
}

File: vm/format/js/src/vmclass.ts
import { VMType } from "./vmtype.js";

export interface SlotInfo {
    readonly parameterTypes: readonly VMType[];
    readonly returnType: VMType;
    readonly symbol: symbol;
}

export interface SlotKey {
    classId: bigint;
    index: bigint;
}

export interface FieldInfo {
    readonly fieldType: VMType;
    readonly symbol: symbol;
}

export interface VMClass {
    readonly prototypeObject: object;

    readonly baseClass: VMClass | null;
    readonly fields: readonly FieldInfo[];
    readonly slots: readonly SlotInfo[];
}
